requests-for-comments/2024-07-01-alerting.md

```diff
@@ -0,0 +1,160 @@
+### Request for comments (RFC): Threshold-based alerting
+
+#### Introduction
+
+This RFC proposes an architecture for an alerting feature in PostHog, which allows users to define conditions that
+trigger alerts based on their data. The system will evaluate these conditions at regular intervals and notify users
+when conditions are met. The design includes detailed logging for debugging and auditing, and efficient handling of
+task execution using Celery.
+
+#### Goals
+
+1. **Condition Evaluation**: Define and evaluate alert conditions at specified intervals.
+2. **Notification System**: Notify users when alert conditions are met.
+3. **Logging and Auditing**: Store detailed logs of each alert check.
+4. **Efficiency and Scalability**: Handle tasks efficiently to reduce load on the system.
+
+#### Architecture
+
+1. **Models**:
+    - **Alert**: Represents the alert configuration.
+    - **AlertCheck**: Logs each evaluation of an alert, including results and notification statuses.
+
+2. **Task Execution**:
+    - Use Celery to handle periodic checks and notifications.
+    - Implement concurrency limits and task chaining for efficient processing.
+
+#### Models
+
+##### Alert Model
+
+```python
+class Alert(models.Model):
+    team = models.ForeignKey("Team", on_delete=models.CASCADE)
+    insight = models.ForeignKey("posthog.Insight", on_delete=models.CASCADE)
+    name = models.CharField(max_length=100)
+    target_value = models.TextField()
+    anomaly_condition = models.JSONField(default=dict)
+```
+
+Credits: @nikitaevg in https://github.com/PostHog/posthog/pull/22554
+
+This main model is responsible for storing alert configurations. We may consider adding fields for alert frequency,
+notification preferences, and other settings based on user requirements. Future integration with CDP is also tbd.
+
+##### AlertCheck Model
+
+The `AlertCheck` model logs each evaluation of an alert:
+- `alert`: Foreign key to the Alert model.
+- `created`: Time of the check.
+- `calculated_value`: Result of the check.
+- `anomaly_condition`: Copy of the condition for the alert in case original got updated.
+- `threshold_met`: Whether the condition was met.
```

Comment:
maybe `trigger_condition` and `trigger_condition_met` ??

Reply:
> trigger_condition

:shrug: sounds good

> trigger_condition_met

I'd prefer having an enum here, IMO it will need more values than 2. You think check_result is not good?

Reply:
Regarding "anomaly": Is there any background or reasoning for this term? I think it evokes other expectations (anomaly detection) and is not quite what we have here with strict set thresholds. I would actually vote to use another term.

Reply:
>On the second thought, maybe it shouldn't be a boolean, but an enum. If there's an error in calculation, I don't know what boolean value can be assigned here. 

We need to decide whether to send out a notification (= threshold met) yes or no. That is boolean.
Errors can be stored separately (see `error_message`).

Reply:
> Is there any background or reasoning for this term?

It just popped in my head. What term would you prefer? I would say a value that is above or below a user-defined threshold _is_ an anomaly.

> We need to decide whether to send out a notification (= threshold met) yes or no. 

1. What if there's a problem with the alert or insight _configuration_, for example the user misconfigured something, we should notify users about that, no? So whether we send out a notification might depend not only on meeting the threshold.
2. If there's an error with the calculation, what value would the "threshold_met" have? I would say it's neither true nor false, and I wouldn't like to have an optional boolean as it doesn't help readability.
3. In the future alerts might get other anomaly detection mechanisms, so it won't be only about thresholds. That's why I think it's better not mention "threshold" here.
4. Enums are cheap, IMO provide better readability, and are extensible, why do you think it's better to use a boolean? 

Sorry, missed this comment last time.

Reply:
i've built a couple of similar matchers in the past and they always evolve to support `&&` and `||` and i always ended up returning something like `{matches: bool, explanation: string }` - so familiarity with something similar is altering my opinion here (doesn't mean i'm right 😊)

i'd definitely have `error` as a separate case, if there's an error the condition was neither met nor not met

in other parts of posthog we ended up having `error_count` and `last_error_date` so that we could decide whether to carry on trying something

----
tbh there's a predicate condition and a predicate result

maybe it's better to have it that generic
----
this might not always be an anomaly to the user - "tell me when an order value is about $1k" is a happy thing

Reply:
but i should leave you @webjunkie to it 😊

Reply:
> if there's an error the condition was neither met nor not met

That worries me a bit, I usually find difficult to interpret optional booleans. My rule of thumb is that if something is slightly more difficult than yes/no then I default to an enum - it's more explicit and more extensible. I'm not advocating for different error types in the check_result. I'm just suggesting we use an enum to express this absence of the outcome more explicitly, have a required enum with three states.

And still "Enums are cheap, IMO provide better readability, and are extensible". It's better to have a door open to extensions, especially with databases, isn't it?

Reply:
One other thing that comes to mind, which might influence the modeling here: We currently would alert each time we check, if the threshold is continued to be breached, right?
If we find the threshold is met, we might check previous `AlertCheck`s to see if this was already the cases there and if a notification has been sent already.

Regarding the boolean discussion: I'm happy to talk about a specific alternative if you may present one.

>this might not always be an anomaly to the user - "tell me when an order value is about $1k" is a happy thing

This was also my thinking.. anomaly is a too strong word, and we can leave it more generic.. But naming is hard 🤷🏻